src/core/runner.js

```diff
@@ -0,0 +1,377 @@
+      throw new Error('No dialogue supplied');
+    }
+    let nodes;
+    if (typeof dialogue === 'string') {
+      nodes = convertYarn(dialogue);
+    } else {
+      nodes = dialogue;
+    }
+
+    nodes.forEach((node) => {
+      if (!node.title) {
+        throw new Error(`Node needs a title: ${JSON.stringify(node)}`);
+      } else if (node.title.split('.').length > 1) {
+        throw new Error(`Node title cannot contain a dot: ${node.title}`);
+      }
+      if (!node.body) {
+        throw new Error(`Node needs a body: ${JSON.stringify(node)}`);
+      }
+      if (this.yarnNodes[node.title]) {
+        throw new Error(`Duplicate node title: ${node.title}`);
+      }
+      this.yarnNodes[node.title] = node;
+    });
+
+    parser.yy.areDeclarationsHandled = false;
+    parser.yy.declarations = {};
+    this.handleDeclarations(nodes);
+    parser.yy.areDeclarationsHandled = true;
+  }
+
+  /**
+   * Set a new variable storage object
+   * This must simply contain a 'get(name)' and 'set(name, value)' function
+   *
+   * Calling this function will clear any existing variable's values
+   */
+  setVariableStorage(storage) {
+    if (typeof storage.set !== 'function' || typeof storage.get !== 'function') {
+      throw new Error('Variable Storage object must contain both a "set" and "get" function');
+    }
+
+    this.variables = storage;
+  }
+
+  /**
+   * Scans for <<declare>> commands and sets initial variable values
+   * @param {any[]} yarn dialogue as string or array
+   */
+  handleDeclarations(nodes) {
+    const exampleValues = {
+      Number: 0,
+      String: '',
+      Boolean: false,
+    };
+
+    const allLines = nodes.reduce((acc, node) => {
+      const nodeLines = node.body.split(/\r?\n+/);
+      return [...acc, ...nodeLines];
+    }, []);
+
+    const declareLines = allLines.reduce((acc, line) => {
+      const match = line.match(/^<<declare .+>>/);
+      return match
+        ? [...acc, line]
+        : acc;
+    }, []);
+    if (declareLines.length) {
+      parser.parse(declareLines.join('\n'));
+    }
+
+    Object.entries(parser.yy.declarations)
+      .forEach(([variableName, { expression, explicitType }]) => {
+        const value = this.evaluateExpressionOrLiteral(expression);
+
+        if (explicitType && typeof value !== typeof exampleValues[explicitType]) {
+          throw new Error(`Cannot declare value ${value} as type ${explicitType} for variable ${variableName}`);
+        }
+
+        if (!this.variables.get(variableName)) {
+          this.variables.set(variableName, value);
+        }
+      });
+  }
+
+  registerFunction(name, func) {
+    if (typeof func !== 'function') {
+      throw new Error('Registered function must be...well...a function');
+    }
+
+    this.functions[name] = func;
+  }
+
+  /**
+   * Generator to return each sequential dialog result starting from the given node
+   * @param {string} [startNode] - The name of the yarn node to begin at
+   */
+  * run(startNode) {
```

Comment:
see this commit for structural changes to runner

https://github.com/mnbroatch/bondage.js/commit/9449905be57370c90bb64ab2c0a878e7f0d5e39c

The old way had unbounded recursion, which caused a stack overflow during some ad-hoc performance testing (quickly jumping between two nodes over and over again). My WSL node stack crapped out at ~4000 jumps.